File: examples/stream.py
import os
import sys
import json
import yaml
import requests
from pathlib import Path
from dotenv import load_dotenv

load_dotenv()

ROOT = Path(__file__).parent.parent
sys.path.insert(0, str(ROOT))

from providers.utils import openai_client

with open(ROOT / "llm.yaml") as f:
    config = yaml.safe_load(f)

provider = config["provider"]
model = config["model"]

print(f"llmkit | {provider} / {model} | streaming")
prompt = input("You: ").strip()
if not prompt:
    sys.exit(0)

print("AI: ", end="", flush=True)

if provider == "local":
    try:
        res = requests.post(
            "http://localhost:11434/api/generate",
            json={"model": model, "prompt": prompt, "stream": True},
            stream=True, timeout=120
        )
        for line in res.iter_lines():
            if line:
                data = json.loads(line)
                print(data.get("response", ""), end="", flush=True)
                if data.get("done"):
                    break
    except requests.exceptions.ConnectionError:
        print("\nError: Ollama is not running. Start it with: ollama serve")

elif provider == "anthropic":
    try:
        import anthropic
        client = anthropic.Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))
        with client.messages.stream(
            model=model, max_tokens=1024,
            messages=[{"role": "user", "content": prompt}]
        ) as stream:
            for text in stream.text_stream:
                print(text, end="", flush=True)
    except Exception as e:
        print(f"\nError: {e}")

else:
    try:
        client = openai_client(provider)
        stream = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}],
            stream=True
        )
        for chunk in stream:
            content = chunk.choices[0].delta.content
            if content:
                print(content, end="", flush=True)
    except Exception as e:
        print(f"\nError: {e}")

print("\n")

File: examples/stream.js
const fs = require("fs");
const readline = require("readline");
const yaml = require("js-yaml");
require("dotenv").config();

const config = yaml.load(fs.readFileSync("llm.yaml", "utf8"));

const API_ENDPOINTS = {
    openai:   "https://api.openai.com/v1/chat/completions",
    groq:     "https://api.groq.com/openai/v1/chat/completions",
    deepseek: "https://api.deepseek.com/v1/chat/completions",
    mistral:  "https://api.mistral.ai/v1/chat/completions",
    together: "https://api.together.xyz/v1/chat/completions",
};

const API_KEYS = {
    openai:   process.env.OPENAI_API_KEY,
    groq:     process.env.GROQ_API_KEY,
    deepseek: process.env.DEEPSEEK_API_KEY,
    mistral:  process.env.MISTRAL_API_KEY,
    together: process.env.TOGETHER_API_KEY,
};

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

console.log(`llmkit | ${config.provider} / ${config.model} | streaming`);

rl.question("You: ", async (prompt) => {
    process.stdout.write("AI: ");

    if (config.provider === "local") {
        const res = await fetch("http://localhost:11434/api/generate", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ model: config.model, prompt, stream: true }),
        });
        const reader = res.body.getReader();
        const decoder = new TextDecoder();
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            const lines = decoder.decode(value).split("\n").filter(Boolean);
            for (const line of lines) {
                const data = JSON.parse(line);
                if (data.response) process.stdout.write(data.response);
            }
        }
    } else {
        const res = await fetch(API_ENDPOINTS[config.provider], {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${API_KEYS[config.provider]}`,
            },
            body: JSON.stringify({
                model: config.model,
                messages: [{ role: "user", content: prompt }],
                stream: true,
            }),
        });
        const reader = res.body.getReader();
        const decoder = new TextDecoder();
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            const lines = decoder.decode(value)
                .split("\n")
                .filter(l => l.startsWith("data: ") && l !== "data: [DONE]");
            for (const line of lines) {
                const data = JSON.parse(line.slice(6));
                const content = data.choices?.[0]?.delta?.content;
                if (content) process.stdout.write(content);
            }
        }
    }

    console.log("\n");
    rl.close();
});
